### ISSUE TITLE
wkwebview Click event is ignored

### ISSUE THREAD

**Issue by ammmmmmmmmmmmm** [flutter/flutter#94325]
When I use  UiKitView wrap Native WKWebview, Quickly click the button from time to time, the second click event works。

The seems like `FlutterTouchInterceptingView`  prevented delivery event of second click event 

this is my demo:
 https://github.com/ammmmmmmmmmmmm/flutter_test_webview.git

<details>
<summary>flutter doctor -v</summary>
[✓] Flutter (Channel stable, 2.5.3, on macOS 11.6 20G165 darwin-x64, locale zh-Hans-CN)
    • Flutter version 2.5.3 at /Users/xuejian/dev/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 18116933e7 (6 weeks ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/xuejian/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 13.0, Build version 13A233
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.62.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.28.0

[✓] Connected device (3 available)
    • iPhone (mobile) • 00008055-001984DE3C43802E • ios            • iOS 15.1 19B74
    • macOS (desktop)     • macos                     • darwin-x64     • macOS 11.6 20G165 darwin-x64
    • Chrome (web)        • chrome                    • web-javascript • Google Chrome 96.0.4664.55
</details>

**Comment by darshankawar** [comment]
@ammmmmmmmmmmmm 
The github repo link is empty. Please provide complete minimal reproducible code sample or repo that we can use and steps to replicate the behavior.
Thanks.

**Comment by ammmmmmmmmmmmm** [comment]
Sorry, I has pushed  my code to `main` branch now. that only run on ios, because this problem will only appear on iOS

**Comment by darshankawar** [comment]
@ammmmmmmmmmmmm 
Please also provide steps to replicate the behavior.

**Comment by ammmmmmmmmmmmm** [comment]
Step: 1
flutter pub get
Step: 2
flutter run
Setp: 3
Click the button twice quickly

Scrren:
The number will change once.

expected： 

The number will change twice.

**Comment by ammmmmmmmmmmmm** [comment]
https://user-images.githubusercontent.com/16349702/144000132-4abd3442-865e-44a5-835a-d885d749c8a0.MP4

**Comment by ammmmmmmmmmmmm** [comment]
In the video, I actually clicked the button twice quickly，But the numbers only changed once

**Comment by lucashsilva** [comment]
I have the same problem.

**Comment by darshankawar** [comment]
@ammmmmmmmmmmmm 
I cloned the repo but unable to run it, as I am getting below compile error:

```
Xcode's output:
↳
    note: Using new build system
    note: Building targets in parallel
    note: Planning build
    note: Constructing build description
    warning: Unable to read contents of XCFileList '/Target Support
    Files/Pods-Runner/Pods-Runner-frameworks-Debug-output-files.xcfilelist' (in target 'Runner' from
    project 'Runner')
    error: Unable to load contents of file list: '/Target Support
    Files/Pods-Runner/Pods-Runner-frameworks-Debug-input-files.xcfilelist' (in target 'Runner' from
    project 'Runner')
    error: Unable to load contents of file list: '/Target Support
    Files/Pods-Runner/Pods-Runner-frameworks-Debug-output-files.xcfilelist' (in target 'Runner' from
    project 'Runner')

Could not build the application for the simulator.
Error launching application on iPhone 12 Pro Max.

```

**Comment by ammmmmmmmmmmmm** [comment]
@darshankawar   The seem like `cocopods` error; try delete Podfile.lock and run `pod install`

**Comment by darshankawar** [comment]
Tried the repo provided and ran it on ios device and noted the same behavior, ie, double tapping on the button increments the counter once, but am not sure if this is expected behavior or a bug. Keeping it open for further insights from the team.

Steps to replicate:

1. Clone the repo.
2. flutter pub get
3. cd ios
4. pod install
5. cd ..
6. flutter run
7. Double click on button to see counter incrementing once.

**Comment by ammmmmmmmmmmmm** [comment]
@darshankawar Thank you!

**Comment by haoqiangyu** [comment]
same problem

**Comment by haoqiangyu** [comment]
This patch (https://gist.github.com/Voxar/5dbef0b76cc8cb55d2fe74a492d41c54) worked for me.

**Comment by jmagman** [comment]
@cyanglaz is this a duplicate of https://github.com/flutter/flutter/issues/61991?

**Comment by haoqiangyu** [comment]
you can also remove some ```gesturerecongizers``` from ```webview.scrolleview.subviews```.
like:
```    
    Class UITapGestureRecognizerClass = NSClassFromString(@"UITapGestureRecognizer");
    Class UITapAndAHalfRecognizer = NSClassFromString(@"UITapAndAHalfRecognizer");
    for (UIView* subView in self.scrollView.subviews) {
        for(UIGestureRecognizer* recognizer in subView.gestureRecognizers){
            if([recognizer isKindOfClass:UITapGestureRecognizerClass]){
                UITapGestureRecognizer* tapRecoginzer = (UITapGestureRecognizer*)recognizer;
                if(tapRecoginzer.numberOfTapsRequired >= 2){
                    NSLog(@"elon:remove recongnizer %@",recognizer);
                    [subView removeGestureRecognizer:recognizer];
                    continue;
                }
            }
            if([recognizer isKindOfClass:UITapAndAHalfRecognizer]){
                NSLog(@"elon:remove recongnizer %@",recognizer);
                [subView removeGestureRecognizer:recognizer];
                continue;
            }
        }
   }
```
tips: you need call remove gesture recongizers function at ```didFinishNavigation``` and when **keyboard hide** event fired
like:
```
- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
  if (!self.shouldEnableZoom) {
    NSString *source =
        @"var meta = document.createElement('meta');"
        @"meta.name = 'viewport';"
        @"meta.content = 'width=device-width, initial-scale=1.0, maximum-scale=1.0,"
        @"user-scalable=no';"
        @"var head = document.getElementsByTagName('head')[0];head.appendChild(meta);";

    [webView evaluateJavaScript:source completionHandler:nil];
  }
  if([webView isKindOfClass:[FLTWKWebView class]]){
    FLTWKWebView* fltWKWebView = (FLTWKWebView*)webView;
    [fltWKWebView removeSpecialGestureRecognizer];
  }
  [_methodChannel invokeMethod:@"onPageFinished" arguments:@{@"url" : webView.URL.absoluteString}];
}
```
``` 
[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleKeyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];

- (void)handleKeyboardWillHide:(NSNotification *)notification {
    [_webView removeSpecialGestureRecognizer];
}
```

**Comment by swordyy7** [comment]
@haoqiangyu  Thanks,I hava set `DelayingGestureRecognizer.enable=NO`;  like this:
`
- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event {
  
    NSArray *gestures = self.superview.gestureRecognizers;
        
    for (int i = 0; i < gestures.count; i++) {
        UIGestureRecognizer *gest = gestures[i];
        if([gest isKindOfClass: NSClassFromString(@"DelayingGestureRecognizer")]) {
            gest.enabled = NO;
        }
    }
    
   return [super hitTest:point withEvent:event];
}
`

I'm not sure if it will cause other issue, At least it solved my problem for now

**Comment by cyanglaz** [comment]
@swordyy7 If you set `DelayingGestureRecognizer.enable=NO`, you won' be able to prevent click events on the webview when a click event happened on a flutter button covering the webview. In this case, both webview and the flutter button will receive/handle the event. 

If there is no such UI in your app, this is fine. 

May I ask why you are implementing your own WKWebView UIkitView? Have you tried https://pub.dev/packages/webview_flutter?

**Comment by haoqiangyu** [comment]
@cyanglaz 
Same problem with the webview_flutter plugin.

His example shows that the simplest native view can reproduce the problem of platform view on iOS.

I think this is a bug of WKWebView. The double-tap gesture will cause the transition from the gesture recognizer's failed state to possible to have a delay in verifying the double-tap. During this delay, events such as touchstart in the webview are not fired. Sorry for my poor english.

**Comment by swordyy7** [comment]
@cyanglaz Same problem with the webview_flutter plugin.

**Comment by swordyy7** [comment]
@haoqiangyu I create a new `Swift`project and I'm use `wkwebview`. when `double-tap` screen , `js` can can receive twice events

**Comment by haoqiangyu** [comment]
> @haoqiangyu I create a new `Swift`project and I'm use `wkwebview`. when `double-tap` screen , `js` can can receive twice events
> 
> 我用 xcode 创建了一个纯swift 项目，快速点击两次按钮， webview 也会处理两次。 这个应该不是wkwebview的问题

Because your newly created iOS project does not add a new gesture recognizer to wkwebview, and the flutter engine adds two internally, and the flutter engine has a state switch in the gesture recognizer. 
See https://github.com/flutter/engine/blob/main/shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

flutter引擎里面有对platform view的添加单独的手势识别器，而你新建的项目没有。

**Comment by swordyy7** [comment]
对啊，所以问题应该出在flutter的这个手势处理上

**Comment by haoqiangyu** [comment]
> 对啊，所以问题应该出在flutter的这个手势处理上

调试下这个手势处理的代码，我认为用法没错，是手势识别器在从failed状态切换到possible状态时存在的问题

**Comment by haoqiangyu** [comment]
> If you figure out a fix, please consider sending a PR 🙂

@jmagman The above is just a temporary solution as some of the WKWebView's gesture recognizers are removed.

**Comment by timukasr** [comment]
Having same problem with webview_flutter plugin. Is it possible to bypass the issue without patching (changing) Flutter/webview_flutter library code? E.g. through iOS method channel or similar. There are some code examples here, but I'm not too familiar with Objective-C or where/how to use it.

**Comment by swordyy7** [comment]
> Having same problem with webview_flutter plugin. Is it possible to bypass the issue without patching (changing) Flutter/webview_flutter library code? E.g. through iOS method channel or similar. There are some code examples here, but I'm not too familiar with Objective-C or where/how to use it.


fix code at below file:
```
webview_flutter/webview_flutter_wkwebview/ios/Classes/FlutterWebView.m
```

like this:
``` object-c
@implementation FLTWKWebView

- (void)setFrame:(CGRect)frame {
  [super setFrame:frame];
  self.scrollView.contentInset = UIEdgeInsetsZero;
  // We don't want the contentInsets to be adjusted by iOS, flutter should always take control of
  // webview's contentInsets.
  // self.scrollView.contentInset = UIEdgeInsetsZero;
  if (@available(iOS 11, *)) {
    // Above iOS 11, adjust contentInset to compensate the adjustedContentInset so the sum will
    // always be 0.
    if (UIEdgeInsetsEqualToEdgeInsets(self.scrollView.adjustedContentInset, UIEdgeInsetsZero)) {
      return;
    }
    UIEdgeInsets insetToAdjust = self.scrollView.adjustedContentInset;
    self.scrollView.contentInset = UIEdgeInsetsMake(-insetToAdjust.top, -insetToAdjust.left,
                                                    -insetToAdjust.bottom, -insetToAdjust.right);
  }
}

- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event {
    NSLog(@"hit test firing");
    

    NSArray *gestures = self.superview.gestureRecognizers;
    
    NSLog(@"gestures2:%@",gestures);
    
    for (int i = 0; i < gestures.count; i++) {
        UIGestureRecognizer *gest = gestures[i];
        if([gest isKindOfClass: NSClassFromString(@"DelayingGestureRecognizer")]) {
            gest.enabled = NO;
        }
      
    }
    
   return [super hitTest:point withEvent:event];
}


@end


`FLTWKWebview` implementation add 
 `- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event` method 
```
to `disable` *DelayingGestureRecognizer*

**Comment by timukasr** [comment]
It seems that for newer webview_flutter_wkwebview versions, this file must be changed: https://github.com/flutter/packages/blob/main/packages/webview_flutter/webview_flutter_wkwebview/ios/Classes/FWFWebViewHostApi.h

Using code by @swordyy7 and ChatGPT, ended up with this solution so that you don't have to change third-party library code:
Create new file `FWFWebView+HitTest.swift`:
```swift
import Foundation
import ObjectiveC.runtime
import UIKit

extension FWFWebView {
    @objc dynamic func custom_hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        if let gestures = superview?.gestureRecognizers {
            for gesture in gestures {
                if NSStringFromClass(type(of: gesture)) == "DelayingGestureRecognizer" {
                    gesture.isEnabled = false
                }
            }
        }

        return custom_hitTest(point, with: event)
    }

    private static let swizzling: Void = {
        let originalSelector = #selector(hitTest(_:with:))
        let swizzledSelector = #selector(custom_hitTest(_:with:))

        let originalMethod = class_getInstanceMethod(FWFWebView.self, originalSelector)
        let swizzledMethod = class_getInstanceMethod(FWFWebView.self, swizzledSelector)

        let didAddMethod = class_addMethod(
            FWFWebView.self,
            originalSelector,
            method_getImplementation(swizzledMethod!),
            method_getTypeEncoding(swizzledMethod!)
        )

        if didAddMethod {
            class_replaceMethod(
                FWFWebView.self,
                swizzledSelector,
                method_getImplementation(originalMethod!),
                method_getTypeEncoding(originalMethod!)
            )
        } else {
            method_exchangeImplementations(originalMethod!, swizzledMethod!)
        }
    }()

    @objc dynamic class func swizzleHitTest() {
        _ = swizzling
    }
}
```

Add this line to `Runner-Bridging-Header.h`:
```header
#import "FWFWebViewHostApi.h"
```

Then active the swizzle in AppDelegate.swift (or wherever needed):
```swift
FWFWebView.swizzleHitTest()
```

**Comment by timukasr** [comment]
With Flutter 3.27 the Swizzle fix in https://github.com/flutter/flutter/issues/94325#issuecomment-1479543992 does not work anymore. I also tried @andreasgangso recommendation by adding `gestureRecognizersBlockingPolicy:FlutterPlatformViewGestureRecognizersBlockingPolicyWaitUntilTouchesEnded` to local version of [webview_flutter_wkwebview](https://github.com/flutter/packages/blob/05a65b4dc78213f78aa33c41d719b3cccfb30123/packages/webview_flutter/webview_flutter_wkwebview/darwin/webview_flutter_wkwebview/Sources/webview_flutter_wkwebview/WebViewFlutterPlugin.swift#L31) plugin, but it did not have any effect.
Help or ideas are welcome.

**Comment by Karibash** [comment]
The method of setting `gestureRecognizersBlockingPolicy` did not work, so I created the following `GestureRecognizer` to detect double taps.

<details><summary>Details</summary>

```dart
class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final controller = WebViewController()
      ..loadRequest(Uri.parse('https://flutter.dev'));
    return WebViewWidget(
      controller: controller,
      gestureRecognizers: {
        Factory(
          () => _DoubleTapEagerGestureRecognizer(
            onDoubleTap: () {
              print('onDoubleTap');
            },
          ),
        ),
      },
    );
  }
}

class _DoubleTapEagerGestureRecognizer extends OneSequenceGestureRecognizer {
  _DoubleTapEagerGestureRecognizer({
    required this.onDoubleTap,
  });

  final void Function(Offset position) onDoubleTap;

  _DoubleTapTracker? _tracker;

  Timer? _timeoutTimer;

  @override
  String get debugDescription => 'ignore double tap';

  @override
  void didStopTrackingLastPointer(int pointer) {}

  @override
  void handleEvent(PointerEvent event) {
    final tracker = _tracker;
    if (event is PointerDownEvent) {
      if (tracker == null) {
        _track(event);
      } else {
        if (!tracker.isWithinSlop(event, kDoubleTapSlop)) {
          stopTrackingPointer(event.pointer);
        } else if (!tracker.isMinTimeElapsed) {
          _track(event);
        }
      }
    } else if (event is PointerUpEvent) {
      stopTrackingPointer(event.pointer);
      if (tracker != null) {
        if (tracker.didPointerUp) {
          _reset();
          onDoubleTap(event.localPosition);
        } else {
          _startTimeoutTimer();
          tracker.pointerUp();
        }
      }
    } else if (event is PointerMoveEvent) {
      if (tracker != null && !tracker.isWithinSlop(event, kDoubleTapTouchSlop)) {
        _reset();
        stopTrackingPointer(event.pointer);
      }
    } else if (event is PointerCancelEvent) {
      _reset();
      stopTrackingPointer(event.pointer);
    }

    resolve(GestureDisposition.accepted);
  }

  void _track(PointerDownEvent event) {
    _stopTimeoutTimer();
    _tracker = _DoubleTapTracker(event);
  }

  void _reset() {
    _stopTimeoutTimer();
    _tracker = null;
  }

  void _startTimeoutTimer() {
    _timeoutTimer ??= Timer(kDoubleTapTimeout, _reset);
  }

  void _stopTimeoutTimer() {
    if (_timeoutTimer != null) {
      _timeoutTimer!.cancel();
      _timeoutTimer = null;
    }
  }
}

class _DoubleTapTracker {
  _DoubleTapTracker(this.event) {
    Timer(kDoubleTapMinTime, _onMinTimeElapsed);
  }

  final PointerDownEvent event;

  var _isMinTimeElapsed = false;

  bool get isMinTimeElapsed => _isMinTimeElapsed;

  var _didPointerUp = false;

  bool get didPointerUp => _didPointerUp;

  void _onMinTimeElapsed() {
    _isMinTimeElapsed = true;
  }

  bool isWithinSlop(PointerEvent event, double slop) {
    final delta = event.position - this.event.position;
    return delta.distance <= slop;
  }

  void pointerUp() {
    _didPointerUp = true;
  }
}
```

</details>

**Comment by timukasr** [comment]
To use Swizzling approach works with Flutter 3.27, you must use `FlutterDelayingGestureRecognizer` instead of `DelayingGestureRecognizer`:

<details>
<summary>
Code
</summary>

```swift
import Foundation
import ObjectiveC.runtime
import UIKit

extension FWFWebView {
    @objc dynamic func custom_hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        if let gestures = superview?.gestureRecognizers {
            for gesture in gestures {
                if NSStringFromClass(type(of: gesture)) == "FlutterDelayingGestureRecognizer" {
                    gesture.isEnabled = false
                }
            }
        }

        return custom_hitTest(point, with: event)
    }

    private static let swizzling: Void = {
        let originalSelector = #selector(hitTest(_:with:))
        let swizzledSelector = #selector(custom_hitTest(_:with:))

        let originalMethod = class_getInstanceMethod(FWFWebView.self, originalSelector)
        let swizzledMethod = class_getInstanceMethod(FWFWebView.self, swizzledSelector)

        let didAddMethod = class_addMethod(
            FWFWebView.self,
            originalSelector,
            method_getImplementation(swizzledMethod!),
            method_getTypeEncoding(swizzledMethod!)
        )

        if didAddMethod {
            class_replaceMethod(
                FWFWebView.self,
                swizzledSelector,
                method_getImplementation(originalMethod!),
                method_getTypeEncoding(originalMethod!)
            )
        } else {
            method_exchangeImplementations(originalMethod!, swizzledMethod!)
        }
    }()

    @objc dynamic class func swizzleHitTest() {
        _ = swizzling
    }
}
```
</details>

**Comment by mboudreau** [comment]
Did anyone got this working in flutter_inappwebview?

**Comment by matovnikov** [comment]
> Did anyone got this working in flutter_inappwebview?

Hi!
If your app only consists of a webview widget and you don't need Flutter gestures, I found a way to disable them completely.

**File:** `/ios/Classes/SwiftFlutterPlugin.swift`

**Add after import:**
```
extension UIView {
    @objc dynamic func custom_addGestureRecognizer(_ gesture: UIGestureRecognizer) {
        if NSStringFromClass(type(of: gesture)) == "FlutterDelayingGestureRecognizer" {
            // заблокировать добавление
            return
        }
        // вызываем "оригинал" (будет swizzled)
        custom_addGestureRecognizer(gesture)
    }

    private static let swizzlingAddGesture: Void = {
        let originalSelector = #selector(addGestureRecognizer(_:))
        let swizzledSelector = #selector(custom_addGestureRecognizer(_:))

        guard let originalMethod = class_getInstanceMethod(UIView.self, originalSelector),
              let swizzledMethod = class_getInstanceMethod(UIView.self, swizzledSelector) else { return }

        method_exchangeImplementations(originalMethod, swizzledMethod)
    }()

    @objc class func swizzleGestureRecognizerAdd() {
        _ = swizzlingAddGesture
    }
}
```

**And in:** `public static func register(with registrar: FlutterPluginRegistrar) {`
Line `UIView.swizzleGestureRecognizerAdd()`
